enterprise/api/v1/views.py

```diff
@@ -357,43 +356,79 @@ def _revoke_enrollment(self, enrollment_api_client, enterprise_enrollment, cours
             'course_id': course_run_id,
             'mode': audit_mode,
         }
-        try:
-            enrollment_api_client.update_course_enrollment_mode_for_user(
-                username=enterprise_customer_user.username,
-                course_id=course_run_id,
-                mode=audit_mode,
-            )
+
+        if self._has_user_completed_course_run(enterprise_enrollment, course_overview):
             LOGGER.info(
-                'Updated LMS enrollment for User {user} and Enterprise {enterprise} in Course {course_id} '
-                'to Course Mode {mode}.'.format(**log_message_kwargs)
+                'License revocation: not updating enrollment in {course_id} for User {user} '
+                'in Enterprise {enterprise}, course is already complete.'.format(**log_message_kwargs)
             )
-        except Exception as exc:  # pylint: disable=broad-except
-            msg = (
-                'Unable to update LMS enrollment for User {user} and Enterprise {enterprise} in Course {course_id} '
-                'to Course Mode {mode} because: {reason}'.format(
-                    reason=str(exc),
-                    **log_message_kwargs
-                )
-            )
-            LOGGER.error('{msg}: {exc}'.format(msg=msg, exc=exc))
-            raise EnrollmentModificationException(msg)
+            return
 
-        enterprise_enrollment.license.revoke()
+        if CourseMode.mode_for_course(course_run_id, audit_mode):
+            try:
+                enrollment_api_client.update_course_enrollment_mode_for_user(
+                    username=enterprise_customer_user.username,
+                    course_id=course_run_id,
+                    mode=audit_mode,
+                )
+                LOGGER.info(
+                    'License revocation: updated LMS enrollment for User {user} and Enterprise {enterprise} '
+                    'in Course {course_id} to Course Mode {mode}.'.format(**log_message_kwargs)
+                )
+            except Exception as exc:  # pylint: disable=broad-except
+                msg = (
+                    'License revocation: unable to update LMS enrollment for User {user} and Enterprise {enterprise} '
+                    'in Course {course_id} to Course Mode {mode} because: {reason}'.format(
+                        reason=str(exc),
+                        **log_message_kwargs
+                    )
+                )
+                LOGGER.error('{msg}: {exc}'.format(msg=msg, exc=exc))
+                raise EnrollmentModificationException(msg)
```

Comment:
I agree with Adam's comment here. To add on, I wonder if it's worth not unenrolling immediately on _any_ exception. It might be worth handling 429s more gracefully, and maybe some other things too?
Anyway, I don't know how nuanced we care to be with this one already somewhat aggressive end state, so feel free to ignore this